bundler/bundler.go

```diff
@@ -51,48 +66,86 @@ func BundleBytes(bytes []byte, configuration *datamodel.DocumentConfiguration) (
-					}
-				}
-			}
+	// indexes := rolodex.GetIndexes()
+	// for _, idx := range indexes {
+	// 	handleRefs(idx, opts)
+	// }
 
-			if mappedReference != nil && !mappedReference.Circular {
-				sequenced.Node.Content = mappedReference.Node.Content
-				continue
-			}
+	idx := rolodex.GetRootIndex()
+	mappedReferences := idx.GetMappedReferences()
+	sequencedReferences := idx.GetRawReferencesSequenced()
 
-			if mappedReference != nil && mappedReference.Circular {
-				if idx.GetLogger() != nil {
-					idx.GetLogger().Warn("[bundler] skipping circular reference",
-						"ref", sequenced.FullDefinition)
-				}
-			}
+	for _, sequenced := range sequencedReferences {
+		mappedReference := mappedReferences[sequenced.FullDefinition]
+		bundleRefTarget(sequenced, mappedReference, opts)
+	}
+
+	return model.Render()
+}
+
+// TODO: use orderedmap as return value
+func bundleRefTarget(ref, refTarget *index.ReferenceNode, opts BundleOptions) (map[string]*index.ReferenceNode, error) {
+	idx := ref.Index
+	if refTarget == nil {
+		if idx.GetLogger() != nil {
+			idx.GetLogger().Warn("[bundler] skipping unresolved reference",
+				"ref", ref.FullDefinition)
 		}
+		return nil, nil
 	}
 
-	indexes := rolodex.GetIndexes()
-	for _, idx := range indexes {
-		compact(idx, false)
+	if refTarget.Circular {
+		if idx.GetLogger() != nil {
+			idx.GetLogger().Warn("[bundler] skipping circular reference",
+				"ref", ref.FullDefinition)
+		}
+		return nil, nil
 	}
-	compact(rolodex.GetRootIndex(), true)
-	return model.Render()
+
+	switch opts.RelativeRefHandling {
+	case RefHandlingInline:
+		ref.Node.Content = refTarget.Node.Content
+	case RefHandlingCompose:
+		// When composing, we need to update the ref values to point to a local reference. At the
+		// same time we need to track all components referenced by any children of the target, so
+		// that we can include them in the final document.
+		//
+		// One issue we might face is that the name of a target component in any given target
+		// document is the same as that of another component in a different target document or
+		// even the root document.
+
+		// Obtain the target's file's index because we should find child references using that.
+		// Otherwise ExtractRefs will use the ref's index and it's absolute spec path for
+		// the FullPath of any extracted ref targets.
+		targetIndex := idx
+		if targetFile := refTarget.DefinitionFile(); targetFile != "" {
+			targetIndex = idx.GetRolodex().GetFileIndex(targetFile)
+		}
+
+		targetMappedReferences := targetIndex.GetMappedReferences()
+
+		bundledComponents := map[string]*index.ReferenceNode{
+			target
+		}
+
+		childRefs := targetIndex.ExtractRefs(refTarget.Node, refTarget.ParentNode, make([]string, 0), 0, false, "")
+
+		for _, childRef := range childRefs {
+			childRefTarget := targetMappedReferences[childRef.FullDefinition]
+			childComponents, err := bundleRefTarget(childRef, childRefTarget, opts)
+			if err != nil {
+				return nil, err
+			}
+		}
```

Comment:
There are two ways to do this. 

1. Go down low **(the hard, long, but the right way)**. 

This would be to do what you said, to manipulate the model at the low end and build up the document manually by creating high level objects from low level ones, and then adding them to the correct component location.

2. Build high  **(the simpler, faster way, questionable on correctness)**

Just build out the new components using the high level, add them as references (https://pb33f.io/libopenapi/modifying/#using-references) and then render out the doc. 

Why is this questionable? well, it means if we provide this as an API, the only way to populate the low level
is to `RenderAndReload` and it would push another run of the model again. Will it cost anything, some CPU and memory - but not much to be honest. Is it right? not exactly - but will work as long as the use is documented.

Reply:
So the answer to your question is, choose the path you can invest in. If you have the time and the stomach, number 1. It will make for a better API, design and a less costly implementation. 

If you do not have time to spare (like any of us) then I suggestion option 2. It will work, it's not ideal, but it's also not wrong either - just not crisp and not performant.